<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working Locally using anaconda, please uncomment the following code and execute it.
Use the version as per your python version.


In [1]:
!pip install pandas beautifulsoup4 requests html5lib

!pip install yfinance
!pip install bs4
!pip install nbformat
!pip install --upgrade plotly

Traceback (most recent call last):
  File "/usr/local/bin/pip3", line 10, in <module>
  File "/usr/local/lib/python3.12/dist-packages/pip/_internal/cli/main.py", line 78, in main
    command = create_command(cmd_name, isolated=("--isolated" in cmd_args))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/pip/_internal/commands/__init__.py", line 114, in create_command
    module = importlib.import_module(module_path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>

In [2]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [ ]:
import plotly.io as pio
pio.renderers.default = "iframe"

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [ ]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. **You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.**


In [ ]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021-06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()
    from IPython.display import display, HTML
    fig_html = fig.to_html()
    display(HTML(fig_html))

Use the make_graph function that we’ve already defined. You’ll need to invoke it in questions 5 and 6 to display the graphs and create the dashboard.
> **Note: You don’t need to redefine the function for plotting graphs anywhere else in this notebook; just use the existing function.**


## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [ ]:
import yfinance as yf

# Extract Tesla stock data
tesla_stock = yf.download('TSLA', start='2010-01-01', end='2021-06-30')

# Display the first few rows of the data
tesla_stock.head()


[*********************100%***********************]  1 of 1 completed


Price,Close,High,Low,Open,Volume
Ticker,TSLA,TSLA,TSLA,TSLA,TSLA
Date,,,,,
2010-06-29,1.592667,1.666667,1.169333,1.266667,281494500
2010-06-30,1.588667,2.028000,1.553333,1.719333,257806500
2010-07-01,1.464000,1.728000,1.351333,1.666667,123282000
2010-07-02,1.280000,1.540000,1.247333,1.533333,77097000
2010-07-06,1.074000,1.333333,1.055333,1.333333,103003500


Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [ ]:
import pandas as pd
from IPython.display import display

REVENUE_URL = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

def get_quarterly_revenue(url: str, match_text: str) -> pd.DataFrame:
    """
    Returns a clean DataFrame with columns ['Date', 'Revenue'] from the course revenue page.
    Handles varying headers and avoids invalid escape sequences.
    """
    # Read the specific table by matching its heading text
    tables = pd.read_html(url, match=match_text, flavor="bs4")
    if not tables:
        raise RuntimeError(f"No table matched: {match_text}")

    raw = tables[0].copy()

    # Auto-pick the first two columns (header names can vary)
    df = raw.iloc[:, :2].copy()

    # Standardize column names to avoid KeyError
    df.columns = ["Date", "Revenue"]

    # Drop any header-like rows that got parsed as data (e.g., if 'Revenue' text leaked into rows)
    # Keep only rows where Revenue contains at least one digit
    df = df[df["Revenue"].astype(str).str.contains(r"\d", na=False)].copy()

    # Clean revenue: remove $ and commas (no invalid escapes)
    df["Revenue"] = (
        df["Revenue"]
        .astype(str)
        .str.replace(r"[$,]", "", regex=True)  # <- avoids SyntaxWarning
        .str.strip()
    )

    # Convert types
    df["Date"] = pd.to_datetime(df["Date"], errors="coerce")
    df["Revenue"] = pd.to_numeric(df["Revenue"], errors="coerce")

    # Final clean
    df = df.dropna(subset=["Date", "Revenue"]).reset_index(drop=True)
    return df

# --- Q2: Tesla ---
tesla_revenue = get_quarterly_revenue(REVENUE_URL, "Tesla Quarterly Revenue")
display(tesla_revenue.head())

# --- Q4: GameStop ---
gme_revenue = get_quarterly_revenue(REVENUE_URL, "GameStop Quarterly Revenue")
display(gme_revenue.head())


,Date,Revenue
0,2022-09-30,21454
1,2022-06-30,16934
2,2022-03-31,18756
3,2021-12-31,17719
4,2021-09-30,13757


ValueError: No tables found matching pattern 'GameStop Quarterly Revenue'

Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Quarterly Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Step-by-step instructions</summary>

```

Here are the step-by-step instructions:

1. Create an Empty DataFrame
2. Find the Relevant Table
3. Check for the Tesla Quarterly Revenue Table
4. Iterate Through Rows in the Table Body
5. Extract Data from Columns
6. Append Data to the DataFrame

```
</details>


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1

We are focusing on quarterly revenue in the lab.
```

</details>


Execute the following line to remove the comma and dollar sign from the `Revenue` column.


In [ ]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"",regex=True)

Execute the following lines to remove an null or empty strings in the Revenue column.


In [ ]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [ ]:
# Extract GameStop stock data
gme_stock = yf.download('GME', start='2010-01-01', end='2021-06-30')

# Reset the index
gme_stock.reset_index(inplace=True)

# Display the first few rows of the data
gme_stock.head()


[*********************100%***********************]  1 of 1 completed


Price,Date,Close,High,Low,Open,Volume
Ticker,,GME,GME,GME,GME,GME
0,2010-01-04,3.854643,3.863060,3.703151,3.714933,26702800
1,2010-01-05,3.959005,3.996036,3.854643,3.856326,21269600
2,2010-01-06,4.044850,4.056633,3.908506,3.948904,21471200
3,2010-01-07,3.443930,3.563441,3.268872,3.368184,164761200
4,2010-01-08,3.415316,3.573541,3.405216,3.474230,47872400


Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data_2`.


In [ ]:
# Q4 — Extract GameStop Quarterly Revenue (robust) + clean + show tail

import pandas as pd
import requests
from bs4 import BeautifulSoup

# If you already defined html_data_2 above, we'll reuse it; otherwise fetch it now.
try:
    html_data_2
except NameError:
    url2 = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
    html_data_2 = requests.get(url2).text

def get_gme_quarterly_revenue(html_stock: str) -> pd.DataFrame:
    """
    Try to read 'GameStop Quarterly Revenue' from stock.html first.
    If not present in your course version, fall back to revenue.htm.
    Then standardize to columns ['Date', 'Revenue'] so the lab cleaning lines work.
    """
    # Try stock.html first
    try:
        tables = pd.read_html(html_stock, match="GameStop Quarterly Revenue", flavor="bs4")
        df = tables[0]
    except ValueError:
        # Fallback: use the course revenue page (same one as Q2)
        rev_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
        df = pd.read_html(rev_url, match="GameStop Quarterly Revenue", flavor="bs4")[0]

    # Keep the first two columns (date + revenue-like), then standardize names
    df = df.iloc[:, :2].copy()
    df.columns = ["Date", "Revenue"]

    # Drop any non-data rows (ensure revenue column has at least one digit)
    df = df[df["Revenue"].astype(str).str.contains(r"\d", na=False)].copy()

    # ---- Run the lab's exact cleaning steps (use raw string to avoid the warning) ----
    df["Revenue"] = df["Revenue"].str.replace(r',|\$', "", regex=True)  # <- no SyntaxWarning
    df.dropna(inplace=True)
    df = df[df["Revenue"] != ""]

    return df.reset_index(drop=True)

# Build gme_revenue and show the last 5 rows
gme_revenue = get_gme_quarterly_revenue(html_data_2)
gme_revenue.tail()


,Date,Revenue
57,2006-01-31,1667
58,2005-10-31,534
59,2005-07-31,416
60,2005-04-30,475
61,2005-01-31,709


Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Quarterly Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column.


> **Note: Use the method similar to what you did in question 2.**  


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`.

```
    
</details>


In [ ]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# URL of the page containing Tesla's revenue data
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

# Send a GET request to the website
response = requests.get(url)

# Parse the page content using BeautifulSoup
soup = BeautifulSoup(response.text, 'html.parser')

# Look for the specific table using its class or ID
table = soup.find('table', {'class': 'historical_data_table'})

# Check if table exists
if table:
    # Parse the table into a DataFrame
    df = pd.read_html(str(table))[0]

    # Display the first few rows of the DataFrame to verify data
    print(df.head())
else:
    print("Table not found")


   Tesla Annual Revenue (Millions of US $)  \
0                                     2021   
1                                     2020   
2                                     2019   
3                                     2018   
4                                     2017   

  Tesla Annual Revenue (Millions of US $).1  
0                                   $53,823  
1                                   $31,536  
2                                   $24,578  
3                                   $21,461  
4                                   $11,759  


## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [ ]:
import yfinance as yf
import pandas as pd

# Download GameStop stock data (let's say for the past 5 years)
gme = yf.Ticker('GME')
gme_data = gme.history(period="5y")

# Show the stock data (you can visualize it or analyze it as needed)
print(gme_data.head())


                             Open    High     Low   Close    Volume  \
Date                                                                  
2020-10-19 00:00:00-04:00  3.3600  3.6250  3.3450  3.4775  52676400   
2020-10-20 00:00:00-04:00  3.5075  3.5350  3.4175  3.4650  26416000   
2020-10-21 00:00:00-04:00  3.4750  3.6050  3.4500  3.5250  21447600   
2020-10-22 00:00:00-04:00  3.5500  3.9675  3.5475  3.7275  64848800   
2020-10-23 00:00:00-04:00  3.7625  3.8450  3.6375  3.7500  26029200   

                           Dividends  Stock Splits  
Date                                                
2020-10-19 00:00:00-04:00        0.0           0.0  
2020-10-20 00:00:00-04:00        0.0           0.0  
2020-10-21 00:00:00-04:00        0.0           0.0  
2020-10-22 00:00:00-04:00        0.0           0.0  
2020-10-23 00:00:00-04:00        0.0           0.0  


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`

```
    
</details>


In [4]:
# Install necessary libraries if not already installed
!pip install yfinance plotly

import yfinance as yf
import pandas as pd
import plotly.graph_objects as go

# Step 1: Extract GameStop Stock Data using yfinance
gme = yf.Ticker('GME')
gme_data = gme.history(period="5y")

# Step 2: Extract GameStop Annual Revenue Data (you can update this if you have different values)
annual_revenue = {
    'Year': [2025, 2024, 2023, 2022, 2021, 2020, 2019, 2018, 2017, 2016, 2015, 2014, 2013, 2012, 2011],
    'Revenue (Millions USD)': [3823, 5273, 5927, 6011, 5090, 6466, 8285, 8547, 7965, 9364, 9296, 9040, 8887, 9551, 9474]
}

# Convert to DataFrame
gme_revenue = pd.DataFrame(annual_revenue)
gme_revenue['Year'] = pd.to_datetime(gme_revenue['Year'], format='%Y')

# Step 3: Resample GameStop Stock Data and Merge with Revenue Data
# Resample the stock data to yearly data (get last stock price for each year)
gme_yearly = gme_data.resample('Y').last()

# Create a 'Year' column for merging
gme_yearly['Year'] = gme_yearly.index.year

# Convert 'Year' in gme_revenue to int
gme_revenue['Year'] = gme_revenue['Year'].dt.year

# Merge the stock price data with the annual revenue data
merged_data = pd.merge(gme_yearly[['Close', 'Year']], gme_revenue[['Year', 'Revenue (Millions USD)']], on='Year', how='inner')

# Step 4: Create the Dashboard using Plotly
fig = go.Figure()

# Plot the GameStop stock price data as a line graph
fig.add_trace(go.Scatter(x=merged_data['Year'], y=merged_data['Close'], mode='lines', name='Stock Price (USD)', line=dict(color='blue')))

# Plot the GameStop annual revenue as a bar chart
fig.add_trace(go.Bar(x=merged_data['Year'], y=merged_data['Revenue (Millions USD)'], name='Revenue (Millions USD)', marker_color='orange'))

# Customize the layout
fig.update_layout(
    title="GameStop Stock and Annual Revenue Dashboard",
    xaxis_title="Year",
    yaxis_title="Stock Price (USD)",
    barmode='group',  # Group the bars for side-by-side comparison
    template='plotly_dark',
    legend_title="Metrics",
    xaxis=dict(
        tickmode='array',
        tickvals=merged_data['Year'],
        ticktext=[str(year) for year in merged_data['Year']]
    ),
)

# Show the interactive plot
fig.show()


/tmp/ipython-input-783267559.py:24: FutureWarning:

'Y' is deprecated and will be removed in a future version, please use 'YE' instead.



<h2>About the Authors:</h2>

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
